stm32: introduce dma v3 driver

youssefz24 opened the pull request:
Hello there, :tophat: 

Edit: my wording isn't helping so I wanted to clear out things so no misunderstandings occur. 
 
This PR adds the stm32_dma_v3 driver and support for the HPDMA (using the V3 driver) on the STM32MP257F-EV1 board. 

It is inspired from [linux version](https://github.com/STMicroelectronics/linux/blob/v6.6-stm32mp/drivers/dma/stm32/stm32-dma3.c) of this driver while maintaining the Zephyr style. (this means that I went through the linux code to understand specifities of the HPDMA, mainly the concept of linked lists that I wasn't familiar with, not that I copied from linux, feel free to compare both logics) 

Up to this point, it covers the following tests/samples: 

- 	samples/boards/st/uart/circular_dma/", // PASS
- 	samples/drivers/uart/async_api/", // PASS
- 	tests/drivers/dma/chan_blen_transfer", // PASS
- 	tests/drivers/dma/chan_link_transfer", // FAIL
- 	tests/drivers/dma/cyclic", // PASS
- 	tests/drivers/dma/loop_transfer", // PASS
- 	tests/drivers/dma/scatter_gather", // PASS

It uses linked lists when needed in a "dynamic" way (it only updates source and destination address). 
The idea is to cover all capabilities of the DMA-V3, some features I am thinking about are: 

- the channel linking (using triggers)
- using the allocated port 1 (depending on conditions) 
- using padding/alignement
- using/preventing packing
- FIFO Mode
- using transfer queues and fragmenting big transfers with linked lists

Any ideas to include these features are welcome. I ll be waiting for reviews and counseling. :tophat: 

Signed-off-by: Youssef Zini [youssef.zini@savoirfairelinux.com](mailto:youssef.zini@savoirfairelinux.com)

youssefz24 committed:
samples/drivers/uart: add stm32mp257f_ev1 overlay

Add stm32mp257f_ev1 overlay for uart async API and circular_dma samples
using uart5 with HPDMA3 channels 0 and 1 for TX and RX respectively.

Signed-off-by: Youssef Zini <youssef.zini@savoirfairelinux.com>

youssefz24 committed:
tests/drivers/dma: add stm32mp257f_ev1 support

Add stm32mp257f_ev1 overlays and related configuration files for the
DMA tests. This includes support for:
* chan_blen_transfer
* cyclic
* loop_transfer
* scatter_gather
Add stm32mp257f_ev1/stm32mp257fxx/m33 to the platform_allow list.

Signed-off-by: Youssef Zini <youssef.zini@savoirfairelinux.com>

youssefz24 committed:
include/zephyr/dt-bindings: add stm32mp2 dma clock

Add the clock bindings for the stm32mp2 dmas, HPDMA(1-3) and LPDMA1.

Signed-off-by: Youssef Zini <youssef.zini@savoirfairelinux.com>

youssefz24 committed:
bindings/dma/st,stm32-dma-v3.yaml: add compatible

Add compatible yaml file for stm32 dma v3 controllers. This is
compatible with the HPDMA and LPDMA controllers in the STM32MP2 series
Socs.

Signed-off-by: Youssef Zini <youssef.zini@savoirfairelinux.com>

youssefz24 committed:
drivers/serial/Kconfig.stm32: add dma v3 support

Add support for the new DMA v3 driver in the STM32 series uart async
API.

Signed-off-by: Youssef Zini <youssef.zini@savoirfairelinux.com>

youssefz24 committed:
drivers/dma/Kconfig.stm32: add dma v3 support

Add dma v3 symbols to Kconfig.stm32 to enable support when dma v3 is
used in the device tree.
Also add the new driver reference in the CMakeLists.txt file.

Signed-off-by: Youssef Zini <youssef.zini@savoirfairelinux.com>

youssefz24 committed:
dts/arm/st/mp2/stm32mp2_m33.dtsi: add hpdma3 node

Add a new HPDMA3 node to the STM32MP2 M33 device tree.
With the current configuration, the Cortex-M33 can only use the HPDMA3
channels 0 to 13.

Signed-off-by: Youssef Zini <youssef.zini@savoirfairelinux.com>

youssefz24 committed:
drivers/dma/dma_stm32_v3.c: add dma v3 driver

Since the STM32MP2 series supports a new DMA peripheral (HPDMA and
LPDMA), this commit introduces a new driver for the STM32 DMA v3.

So far, it includes basic functionality for the HPDMA channels, enabling
the use of async APIs (tested using uart on the STM32MP257F-EV1 board),
as well as support circular dma and loop transfers (reload mode and
suspend/resume).

The driver is designed to be compatible with the existing Zephyr DMA
API, while being inspired heavily by the linux stm32 dma3 driver[1].

The classes and functions used are thought to allow for easy linked list
operations in the future, as well as support scatter gather, cyclic
transfers and channel linking.

[1]: https://github.com/STMicroelectronics/linux/blob/v6.6-stm32mp/drivers/dma/stm32/stm32-dma3.c

Signed-off-by: Youssef Zini <youssef.zini@savoirfairelinux.com>

youssefz24 committed:
boards/st/stm32mp257f_ev1: add dma to board yaml

Add dma as a supported driver in the STM32MP257F-EV1 board YAML file.

Signed-off-by: Youssef Zini <youssef.zini@savoirfairelinux.com>

youssefz24 committed:
drivers/dma/dma_stm32_v3.c: add LLI support

Add support for linked list items (LLI) in the STM32 DMA v3 driver.
This allows for more flexible DMA configurations, such as scatter-gather
and cyclic transfers.
This allows to pass both the dma_m2m_cyclic and the dma_m2m_sg tests on
the stm32mp257f_ev1 board.

Signed-off-by: Youssef Zini <youssef.zini@savoirfairelinux.com>

mathieuchopstm commented on drivers/dma/dma_stm32_v3.c:
```diff
@@ -0,0 +1,763 @@
+ */
+#define CONFIG_DMA_STM32_NUM_DESCRIPTORS_PER_CHANNEL 24
+
+static const uint32_t table_src_size[4] = {
```
Nit: Omit size if it is a magic number
```suggestion
static const uint32_t table_src_size[] = {
```

ditto two arrays below

mathieuchopstm commented on drivers/dma/dma_stm32_v3.c:
```diff
@@ -0,0 +1,763 @@
+	return 0;
+}
+
+inline bool dma_stm32_is_tc_active(const DMA_TypeDef *dma, uint32_t channel)
+{
+	return LL_DMA_IsActiveFlag_TC(dma, channel);
+}
```
1. Don't mark external functions as `inline`.
```suggestion
bool dma_stm32_is_tc_active(const DMA_TypeDef *dma, uint32_t channel)
{
	return LL_DMA_IsActiveFlag_TC(dma, channel);
}
```
Ditto all. (If you want the inlining, move these to the `.h` and make them `static inline` instead)

2. The LL functions should be used directly as these are mere wrappers

youssefz24 replied:
I replaced all "useless" wrappers with direct LL function calls, avoiding the use of inline since LL functions already compile as such.

mathieuchopstm commented on drivers/dma/dma_stm32_v3.c:
```diff
@@ -0,0 +1,763 @@
+	return LL_DMA_IsActiveFlag_USE(dma, channel);
+}
+
+inline bool dma_stm32_is_te_active(const DMA_TypeDef *dma, uint32_t channel)
+{
+	return dma_stm32_is_dte_active(dma, channel) && dma_stm32_is_ule_active(dma, channel) &&
+	       dma_stm32_is_use_active(dma, channel);
+}
```
(FTR, functions like these are fine as they are not mere wrappers - but this should use raw LL)

mathieuchopstm commented on drivers/dma/dma_stm32_v3.h:
```diff
@@ -0,0 +1,84 @@
+	struct stm32_pclken pclken;
+	void (*config_irq)(const struct device *dev);
+	bool support_m2m;
+	uint32_t base;
```
```suggestion
	DMA_TypeDef *base;
```
and remove all casts in driver

mathieuchopstm commented on dts/bindings/dma/st,stm32-dma-v3.yaml:
```diff
@@ -0,0 +1,113 @@
+description: |
+  STM32 DMA controller V3.
+
+  The STM32 DMA3 is a direct memory access controller with different features
+  depending on its hardware configuration.
+  It is either called LPDMA (Low Power), GPDMA (General Purpose) or HPDMA (High
+  Performance).
+
+  Its hardware configuration registers allow to dynamically expose its features.
+
+  GPDMA and HPDMA support 16 independent DMA channels, while only 4 for LPDMA.
+  GPDMA and HPDMA support 256 DMA requests from peripherals (REQSEL[7:0]),
+  32 for LPDMA (REQSEL[4:0]).
```
Kconfig says the driver only supports HPDMA. Binding should not be talking about something else.

youssefz24 replied:
Done

mathieuchopstm commented on dts/bindings/dma/st,stm32-dma-v3.yaml:
```diff
@@ -0,0 +1,113 @@
+  - bit 20: Destination Burst Increment
+            0x0: STM32_DMA_DEST_BURST_NO_INC: no burst increment
+            0x1: STM32_DMA_DEST_BURST_INC: burst increment
+  - bit 21: Destination Allocated Port
+            0x0: STM32_DMA_DEST_PORT_0: Port 0
+            0x1: STM32_DMA_DEST_PORT_1: Port 1
```
AFAICT driver ignores this and always uses `PORT0`. Should be dropped. Ditto all features not supported yet.

youssefz24 replied:
The idea is to support those features that why I'm doing this PR at this point. 
For allocated port both ports can be used but I am still figuring out how to properly condition which to use. This PR is my way of getting information from people who do have the skills and info to do so. ;)

mathieuchopstm replied:
Fair enough.

mathieuchopstm commented on drivers/dma/dma_stm32_v3.c:
```diff
@@ -0,0 +1,763 @@
+		return -ENOTSUP; /* TODO: support different sizes */
+	}
+	width = config->dest_data_size;
+	dest_index = find_lsb_set(config->dest_data_size) - 1;
+	src_index = find_lsb_set(config->source_data_size) - 1;
```
```suggestion
	dest_index = LOG2(config->dest_data_size);
	src_index = LOG2(config->source_data_size);
```

mathieuchopstm reviewed (commented):
(first comments from non-exhaustive review)

youssefz24 commented:
Update, I rebased, reduced the cognitive complexity in the configure function (solved the issue raised by sonarqube) and made some changes following the first comments.

teburd commented on dts/bindings/dma/st,stm32-dma-v3.yaml:
```diff
@@ -0,0 +1,106 @@
+    rx using channel 0 with request 17 for uart5_rx
+    tx using channel 1 with request 18 for uart5_tx
+    uart5 {
+      dmas = <&hpdma3 0 18 (STM32_DMA_PERIPH_TX | STM32_DMA_PRIORITY_HIGH)>,
```
typically these sorts of dma cells have solely a node reference and channel identifier, maybe a handshake peripheral number. The choice to add bit field options here for linked list, cyclic, or various other modes of operation is interesting but I think gives the illusion that perhaps changing these fields will actually change the way the DMA is used at a peripheral driver level.

Is that actually the intent here? To allow say, a uart to work in linked list or cyclic mode, with 16, 32 or 64bit data words? That doesn't seem logical to me.

teburd replied:
That's my point, the uart driver is going to configure the DMA for how it wishes to use it, having these fields here is confusing and would imply to someone configuring their device that these are options that can be changed for say... how the uart driver may use the DMA, but is that really the case? Can I change say how the st uart driver uses DMA to use the linked list descriptors or cyclic buffer mode? That would, I think, be a very different looking driver.

youssefz24 replied:
So what you re asking for is either get the uart driver to actually be able to use Linked Lists,  (I don t see the use of having 32bits transfers as the hardware will only transfer 8 anyways), and remove all the unused parameters from the dts binding ? 

Then again how would you have a custom configuration for Hardware transfer requests ?

teburd replied:
I'm saying don't provide bits in DTS defines that don't make sense from a user configuration perspective, as that's what it is. It's user configurable. If a driver wishes to set word sizing, linked list or cyclic, what the watermark would be, etc etc for a DMA transfer that's entirely up to the *driver* using the peripheral to do this. Otherwise what I'd expect is that a poor user comes along, sees that ST supports linked list transfers or cyclic transfers with their DMA and any peripheral, only to confusingly either break the peripheral driver or be silently ignored. Either of those outcomes is a poor user experience for someone setting up their devicetree.

mathieuchopstm replied:
> That's my point, the uart driver is going to configure the DMA for how it wishes to use it, having these fields here is confusing and would imply to someone configuring their device that these are options that can be changed for say... how the uart driver may use the DMA, but is that really the case? Can I change say how the st uart driver uses DMA to use the linked list descriptors or cyclic buffer mode? That would, I think, be a very different looking driver.

It seems to be the case, actually: DMA configuration `cyclic` depends on DT property
https://github.com/zephyrproject-rtos/zephyr/blob/9135918249b36dbbd675db8c4f94173c147438f3/drivers/serial/uart_stm32.c#L2377-L2378

and the driver checks whether it's enabled or not during its operation:
https://github.com/zephyrproject-rtos/zephyr/blob/9135918249b36dbbd675db8c4f94173c147438f3/drivers/serial/uart_stm32.c#L1162-L1173

Whether all drivers support all use cases is another question but there are at least *some* drivers that consume *some* options to affect their behavior.

(+ I don't remember the exact DMA API, but as far as I understand, whether a DMA transfer is done with a "direct" programming or a linked list should be an implementation detail of the DMA driver?)

youssefz24 replied:
> whether a DMA transfer is done with a "direct" programming or a linked list should be an implementation detail of the DMA driver? 

It seems that it is up to the driver to use one of the two, but that is depending on the usage/configuration. 
The DMA struct from Zephyr enables the use of chained blocks, so one way is saying if you have more than one block to configure linked lists, one can also juste configure a said channel for a transfer and call the `reload` function to do another transfer in direct programming with changing only the addresses, having the same effect with only direct programming. 

So I d say the usage of linked lists or direct programmign depends on how the user side uses the DMA API and its structure.

In my case I tried exposing all the ways depending on the available tests. Si linked lists are used with cyclic and scatter gather tests as they need multiple transfers. and it is better to do so fully with DMA rather than call the CPU at each end of transfer. 

Either way, some of the symbols I added in the yaml are in fact not consumable so far so I guess I ll just remove them for next batch and add them only when that s the case. :D

youssefz24 replied:
@teburd PTAL :p

teburd commented on drivers/dma/dma_stm32_v3.h:
Header file shouldn't be needed here unless the intent is to share these types and functions with other dma driver variants which I don't see happening here.

youssefz24 replied:
Removed the header, as advised. :+1:

teburd commented on drivers/dma/dma_stm32_v3.c:
```diff
@@ -0,0 +1,724 @@
+		return 0;
+	}
+
+	if (!LL_DMA_IsEnabledChannel(dma, channel)) {
```
If the channel is enabled or not, stop is allowable. See the DMA state diagram in driver expectations https://docs.zephyrproject.org/latest/hardware/peripherals/dma.html#driver-implementation-expectations

youssefz24 replied:
The way I see this, the "LL_DMA_IsEnabledChannel" is more like a not running state as it is only changed after a `dma_stm32_start`  call.  

So in this regard, it returns a 0 early when trying to stop before a start was called (just after configuration for example as it is done in the loop_transfer test). Removing this early exit will cause the stop function to actually disable the channel, making it fail to start.

teburd commented on drivers/dma/dma_stm32_v3.c:
```diff
@@ -0,0 +1,724 @@
+	}
+
+	LL_DMA_SuspendChannel(dma, channel);
+	while (timeout-- > 0) {
```
use a WAIT_FOR macro here instead to poll the register status https://docs.zephyrproject.org/latest/doxygen/html/group__sys-util.html#ga68eb35df6b4715714312df90209accee

teburd commented on drivers/dma/dma_stm32_v3.c:
```diff
@@ -0,0 +1,724 @@
+	uint32_t timeout = 10;
+
+	LL_DMA_DisableChannel(dma, channel);
+	while (timeout-- > 0) {
```
WAIT_FOR https://docs.zephyrproject.org/latest/doxygen/html/group__sys-util.html#ga68eb35df6b4715714312df90209accee

youssefz24 replied:
Done

teburd commented on drivers/dma/dma_stm32_v3.c:
```diff
@@ -0,0 +1,724 @@
+
+	LL_DMA_ConfigAddresses(dma, channel, hwdesc.channel_sar, hwdesc.channel_dar);
+
+	/* Enable various interrupts */
```
Is each one of these really necessary or is there a register you can write a bunch of bits simultaneously to?

youssefz24 replied:
They all write different bits of the same register. I am only using them because STM32 tend to use HAL functions to modify bits and registers rather than having a `reg_modify` type function in the driver. 
I d say it is more convenient to understand for someone going through the code. I could wrap them in one `enable_interrupts` functions if necessary ? 
Let me know :tophat:

youssefz24 replied:
It's not difficult that s not an issue. 
I am just saying that the driver uses the STM32 HAL functions, as other drivers from ST do so handling registers directly for DMA would be standing out imo. 

I ll do as you advised for now, so long no one raises a different opinion. 

I m just saying it stands out that's all. :p

PS: I also prefer handling registers directly, but I have to follow the same logic as other drivers. So I avoid to abide by my personal preferences. :c

youssefz24 commented:
> Hi, is this driver compatible with only STM32MP2 series, or it can be used with other STM32 families like STM32H7RS?

Hello @ExaltZephyr 
In theory, the DMA-V3 driver is compatible with LPDMA, GPDMA and HPDMA peripherals as the three of them share the same configuration logic (arbitration and request multiplexer included in the peripheral, linked lists management ...). 
And based on: [STM32H7RS DMA](https://www.st.com/content/ccc/resource/training/technical/product_training/group1/2d/57/e4/3c/e5/34/4d/6d/stm32h7rs-system-dma-overview-dmaovw/files/stm32h7rs-system-dma-overview-dmaovw.pdf/jcr:content/translations/en.stm32h7rs-system-dma-overview-dmaovw.pdf) it seems that it has the same peripherals. 
However as I can't test it on hardware (I only have access to STM32MP257 boards) I can't say for sure. 
Hope this helps. :tophat:

mathieuchopstm commented on include/zephyr/dt-bindings/dma/stm32_dma.h:
```diff
@@ -70,4 +72,56 @@
 
 /** @} */
 
+/**
+ * @name custom HPDMA-specific configuration flags
+ * @{
+ */
+/* HPDMA  Source Burst Increment config on bit 18 */
```
(the driver for now only supports HPDMA, but it can probably be modified later on to work with all DMAv3 variants)

youssefz24 commented on drivers/dma/dma_stm32_v3.c:
```diff
@@ -0,0 +1,757 @@
+	return LL_DMA_IsEnabledIT_HT(dma, channel) && LL_DMA_IsActiveFlag_HT(dma, channel);
+}
+
+static void dma_stm32_disable_it(const DMA_TypeDef *dma, uint32_t channel)
```
While it is true that we are writing to it, what we pass as an argument is its address which is constant. (It is the address from dts that doesn t change `dev_config->base`)

youssefz24 replied:
So I removed the const qualifier from the whole driver. 

> One may argue as to whether the const is "correct" or not, but I don't think this we should even care: it is not needed here so we may as well get rid of it.

The way I see it is that the const qualifier in this context tells that this is a base address, it is constant as it describes the peripheral address in memory, and all the changes that are made are done by offsetting this address so in this sense it is constant. But as you said it is not needed here, and the fact that we are having this conversation shows that having the const qualifier would only be confusing. 

Anyways, let me know if more changes are needed in this regard. :tophat:

youssefz24 commented on drivers/dma/dma_stm32_v3.c:
```diff
@@ -0,0 +1,757 @@
+	/* DMAx CTR2 Configuration */
+	hwdesc.channel_tr2 = 0;
+	hwdesc.channel_tr2 |= ll_direction;
+	hwdesc.channel_tr2 |= config->complete_callback_en ? LL_DMA_TCEM_BLK_TRANSFER
```
In the `LSM_1LINK_EXECUTION` you are refering to, that s actually about wether to use linked lists or not depending on `block_count` and `cyclic` mode, so a call function to configure linked lists was needed. 

Here it s just to configure what the `transfer_complete` event means. Either way, it will be like: 
```c
	if (config-> complete_callback_en == LL_DMA_TCEM_BLK_TRANSFER) {
		hwdesc.channel_tr2 |= LL_DMA_TCEM_BLK_TRANSFER;
	} else {
		hwdesc.channel_tr2 |= LL_DMA_TCEM_EACH_LLITEM_TRANSFER;
	}
```
which is the same as: 
```c
	hwdesc.channel_tr2 |= config->complete_callback_en ? LL_DMA_TCEM_BLK_TRANSFER
							   : LL_DMA_TCEM_EACH_LLITEM_TRANSFER;
 ```

But  yeah I ll change it

youssefz24 commented on drivers/dma/dma_stm32_v3.c:
```diff
@@ -0,0 +1,757 @@
+	return -ETIMEDOUT;
+}
+
+static int dma_stm32_validate_arguments(const struct device *dev, struct dma_config *config)
+{
+	if (dev == NULL || config == NULL) {
+		LOG_ERR("Invalid arguments: dev=%p, config=%p", dev, config);
+		return -EINVAL;
+	}
+
+	return 0;
+}
```
Ow yeah I am refering to the metric of SonarQube uses. In the case of this function it doesn't really change much to wrap the function as the function is not that complicated to begin with. 

The checking `dev != NULL` is indeed unnecessary given that we use the API as intended, but we never know. 
Anyways, I ll move it inline ;)

etienne-lms commented on drivers/dma/dma_stm32_v3.c:
```diff
@@ -0,0 +1,754 @@
+		channel_config->dma_callback(dev, channel_config->user_data, callback_arg,
+					     DMA_STATUS_BLOCK);
+	} else if (dma_stm32_is_tc_irq_active(dma, channel)) {
+		if (channel_config->cyclic == false) {
```
```suggestion
		if (!channel_config->cyclic) {

youssefz24 replied:
done

youssefz24 commented on dts/bindings/dma/st,stm32-dma-v3.yaml:
```diff
@@ -0,0 +1,76 @@
+            0x0: STM32_DMA_PERIPH_8BITS: Byte (8 bits)
+            0x1: STM32_DMA_PERIPH_16BITS: Half-word (16 bits)
+            0x2: STM32_DMA_PERIPH_32BITS: Word (32 bits)
+            0x3: STM32_DMA_PERIPH_64BITS: Double-word (64 bits)
```
done

youssefz24 committed:
include/zephyr: dma stm32: add macro parentheses

Add parentheses to macros in drivers and dt-bindings stm32 dma files to
avoid potential issues with operator precedence when these macros are
used in expressions.

Signed-off-by: Youssef Zini <youssef.zini@savoirfairelinux.com>

ADESTM commented on drivers/dma/dma_stm32_v3.c:
```diff
@@ -0,0 +1,690 @@
+		hwdesc.channel_tr2 |= LL_DMA_TCEM_BLK_TRANSFER;
+	} else {
+		hwdesc.channel_tr2 |= LL_DMA_TCEM_EACH_LLITEM_TRANSFER;
+	}
```
`LL_DMA_TCEM_BLK_TRANSFER` and `LL_DMA_TCEM_EACH_LLITEM_TRANSFER` are rather the same in term of transfer status: none of them indicates the end of the whole transfer.
`LL_DMA_TCEM_BLK_TRANSFER` triggers TC event each time CxBR1.BNDT=0,
`LL_DMA_TCEM_EACH_LLITEM_TRANSFER` triggers TC event each time the next LLI (if any) is loaded after CxBR1.BNDT=0.

ADESTM commented on dts/arm/st/mp2/stm32mp2_m33.dtsi:
```diff
@@ -430,6 +431,19 @@
+				     <77 0>, <78 0>;
+			dma-channels = <14>;
+			dma-requests = <183>;
+			dma-offset = <0>;
```
there is no `status = "disabled";` here. So, why do you need to enable hpdma3 node elsewhere?

etienne-lms replied:
Indeed, hpdam3 node should be default disabled in the SoC DTSI and enabled from board DTS or DTS overlay files.

youssefz24 replied:
I do have it it just got replaced during rebase with the wdg node. :sweat: :sweat:

ADESTM commented on drivers/dma/dma_stm32_v3.c:
```diff
@@ -270,11 +269,69 @@ static int dma_stm32_validate_transfer_sizes(struct dma_config *config)
+	}
+
+	LL_DMA_EnableIT_HT(dma, channel);
+}
```
All blocks have the same length (no CBR1 update) ?

Why do you enable half transfer interrupt ?

avolmat-st commented:
Hello,
jumping a bit late in this conversation.
Maybe I missed an important point, but there is already within Zephyr a driver for the GPDMA of the STM32U5 which can be found in **drivers/dma/dma_stm32u5.c**. Considering that normally LPDMA / GPDMA / HPDMA are very close, can't this driver be used as a starting point to add any potential missing bits ?

etienne-lms commented on drivers/dma/dma_stm32_v3.c:
```diff
@@ -0,0 +1,690 @@
+{
+	mem_addr_t dma_channel_ccr_reg = (mem_addr_t)&dma_stm32_get_channel_addr(dma, channel)->CCR;
+
+	sys_clear_bits(dma_channel_ccr_reg, DMA_CCR_TCIE | DMA_CCR_HTIE | DMA_CCR_USEIE |
+						    DMA_CCR_ULEIE | DMA_CCR_DTEIE | DMA_CCR_SUSPIE);
```
Indentation, remove a tab:
```suggestion
	sys_clear_bits(dma_channel_ccr_reg, DMA_CCR_TCIE | DMA_CCR_HTIE | DMA_CCR_USEIE |
					    DMA_CCR_ULEIE | DMA_CCR_DTEIE | DMA_CCR_SUSPIE);

etienne-lms commented on drivers/dma/dma_stm32_v3.c:
```diff
@@ -0,0 +1,690 @@
+	mem_addr_t dma_channel_cfcr_reg =
+		(mem_addr_t)&dma_stm32_get_channel_addr(dma, channel)->CFCR;
+
+	sys_set_bits(dma_channel_cfcr_reg, DMA_CFCR_TCF | DMA_CFCR_HTF | DMA_CFCR_DTEF |
+						   DMA_CFCR_ULEF | DMA_CFCR_USEF | DMA_CFCR_TOF |
+						   DMA_CFCR_SUSPF);
```
Indentation: remove a tab.

etienne-lms replied:
Strange, indentation is weird here :-(

etienne-lms commented on drivers/dma/dma_stm32_v3.c:
```diff
@@ -0,0 +1,690 @@
+{
+	const struct dma_stm32_config *dev_config = dev->config;
+	struct dma_stm32_channel *channel_config;
+	uint32_t channel = dma_stm32_id_to_channel(id);
```
Check `id < Check `id < id >= dev_config->max_channels`?`?
Ditto in `dma_stm32_stop()`.

etienne-lms commented on drivers/dma/dma_stm32_v3.c:
```diff
@@ -0,0 +1,690 @@
+		COND_CODE_1(DT_NODE_HAS_PROP(__node, clocks),                                      \
+			    (.pclken = { .bus = __bus, .enr = __cenr },),                          \
+			    (/* Nothing if clocks not present */))                                 \
+			.config_irq = dma_stm32_config_irq_##index,                                \
```
Indentation
```suggestion
		.config_irq = dma_stm32_config_irq_##index,                                        \

youssefz24 replied:
Done

etienne-lms commented on drivers/dma/dma_stm32_v3.c:
```diff
@@ -0,0 +1,690 @@
+		.max_channels =                                                                    \
+			DT_INST_PROP_OR(index, dma_channels, DT_NUM_IRQS(DT_DRV_INST(index))),     \
+		.channels = dma_stm32_channels_##index,                                            \
+		.linked_list_buffer = dma_stm32_linked_list_buffer##index};                        \
```
For consistency, prefer:
```suggestion
		.linked_list_buffer = dma_stm32_linked_list_buffer##index,                         \
	};                                                                                         \

youssefz24 replied:
Done

etienne-lms commented on drivers/dma/dma_stm32_v3.c:
```diff
@@ -270,11 +269,69 @@ static int dma_stm32_validate_transfer_sizes(struct dma_config *config)
+	channel = dma_stm32_id_to_channel(id);
+	LL_DMA_SetLinkedListBaseAddr(dma, channel, base_addr);
+
+	for (int i = 0; i < config->block_count; i++) {
```
Prefer use type `uint` since block_count is unsigned. Would prefer some compiler to complain on signed/unsigned comparison.

etienne-lms reviewed (commented):
Some minor comments.

erwango commented:
> Hello,
jumping a bit late in this conversation.
Maybe I missed an important point, but there is already within Zephyr a driver for the GPDMA of the STM32U5 which can be found in drivers/dma/dma_stm32u5.c. Considering that normally LPDMA / GPDMA / HPDMA are very close, can't this driver be used as a starting point to add any potential missing bits ?

Thanks @avolmat-st for raising this point. I agree that rather than duplicating a driver we'd better work on cleaning/completing the existing one.
@youssefz24 Based on the work you made for this PR, would you mind having a check on `drivers/dma/dma_stm32u5.c` and see what would be required to make it work on MP2 ? I'd prefer we work on consolidating the existing and minimize duplication, and I guess there is potential for improvement in many ways (like making it clear which IP are concerned, which obviously we're failing at today)

youssefz24 commented on drivers/dma/dma_stm32_v3.c:
```diff
@@ -0,0 +1,690 @@
+struct dma_stm32_config {
+	struct stm32_pclken pclken;
+	void (*config_irq)(const struct device *dev);
+	bool support_m2m;
```
Done

youssefz24 commented on drivers/dma/dma_stm32_v3.c:
```diff
@@ -0,0 +1,690 @@
+	return channel_nr[id];
+}
+
+static inline DMA_Channel_TypeDef *dma_stm32_get_channel_addr(DMA_TypeDef *dma,
```
removed

youssefz24 commented on drivers/dma/dma_stm32_v3.c:
```diff
@@ -0,0 +1,690 @@
+		LL_DMA_CHANNEL_12, LL_DMA_CHANNEL_13, LL_DMA_CHANNEL_14, LL_DMA_CHANNEL_15,
+	};
+
+	__ASSERT_NO_MSG(id < ARRAY_SIZE(channel_nr));
```
Done

youssefz24 commented on drivers/dma/dma_stm32_v3.c:
```diff
@@ -0,0 +1,690 @@
+#define DMA_STM32_MAX_DATA_ITEMS 0xffff
+
+/* Since at this point we only support cyclic mode , we only need 3 descriptors
+ * at most to update the source and destinantion addresses and the update
```
Done
